File: java/src/test/java/org/taymyr/lagom/javadsl/broker/TestTopicPublisherImpl.java
package org.taymyr.lagom.javadsl.broker;

import akka.NotUsed;
import com.lightbend.lagom.javadsl.api.ServiceCall;

import javax.inject.Inject;

import static akka.NotUsed.notUsed;

class TestTopicPublisherImpl implements TestTopicPublisher {
    private SimpleTopicProducersRegistry simpleTopicProducersRegistry;

    @Inject
    public TestTopicPublisherImpl(TestTopicService testTopicService, SimpleTopicProducersRegistry simpleTopicProducersRegistry) {
        this.simpleTopicProducersRegistry = simpleTopicProducersRegistry.register(testTopicService);
    }

    @Override
    public ServiceCall<NotUsed, NotUsed> publishWithoutKey(String msg) {
        return notUsed -> simpleTopicProducersRegistry.get(TestTopicService.TOPIC_WITHOUT_KEYS).send(msg)
            .thenApply(x -> notUsed());
    }

    @Override
    public ServiceCall<NotUsed, NotUsed> publishWithKey(String msg) {
        return notUsed -> simpleTopicProducersRegistry.get(TestTopicService.TOPIC_WITH_KEYS).send(msg)
            .thenApply(x -> notUsed());
    }

    @Override
    public ServiceCall<NotUsed, NotUsed> enqueueToPublish(String msg) {
        return notUsed -> simpleTopicProducersRegistry.get(TestTopicService.TOPIC_WITH_KEYS).enqueue(msg)
            .thenApply(x -> notUsed());
    }
}

File: java/src/test/kotlin/org/taymyr/lagom/javadsl/broker/TestTopicPublisherImpl.kt
package org.taymyr.lagom.javadsl.broker

import akka.NotUsed
import com.lightbend.lagom.javadsl.api.ServiceCall
import javax.inject.Inject

class TestTopicPublisherImpl @Inject constructor(
    testTopicService: TestTopicService,
    simpleTopicProducersRegistry: SimpleTopicProducersRegistry
) : TestTopicPublisher {
    private val simpleTopicProducersRegistry: SimpleTopicProducersRegistry =
        simpleTopicProducersRegistry.register(testTopicService)

    override fun publishWithoutKey(msg: String): ServiceCall<NotUsed, NotUsed> {
        return ServiceCall {
            simpleTopicProducersRegistry.get(TestTopicService.TOPIC_WITHOUT_KEYS).publish(msg)
                .thenApply { NotUsed.getInstance() }
        }
    }

    override fun publishWithKey(msg: String): ServiceCall<NotUsed, NotUsed> {
        return ServiceCall {
            simpleTopicProducersRegistry.get(TestTopicService.TOPIC_WITH_KEYS).publish(msg)
                .thenApply { NotUsed.getInstance() }
        }
    }
}